### ISSUE TITLE
Default QueryRunner instances for (Maybe (Maybe a))

### ISSUE THREAD

**Issue by bergmark** [karamaan/karamaan-opaleye#61]
If you do a left join on a table that has a Maybe field you will get something like

``` haskell
No instance for (Default QueryRunner (Wire (Maybe (Maybe Int))) (Maybe (Maybe Int)))
```

Two possible implementations:

``` haskell
instance (FromField a, Default QueryRunner (Wire (Maybe a)) (Maybe a)) => Default QueryRunner (Wire (Maybe (Maybe a))) (Maybe (Maybe a)) where
  def = fieldQueryRunner
```

``` haskell
instance (FromField a, Default QueryRunner (Wire a) a) => Default QueryRunner (Wire (Maybe a)) (Maybe a) where
  def = fieldQueryRunner
```

Are both valid? The second one overlaps with what `RunQuery.TH` generates, is it too general or can we remove the TH generated instances? 

I also don't understand why adding the second instance to `RunQuery` doesn't give an overlap error inside that module. I assumed we could remove all the `(Wire (Maybe X)) (Maybe X)` instances.

**Comment by tomjaguarpaw** [comment]
Yeah this is an annoyance of the null semantics of SQL.  Note that there is no distinction in SQL between `Wire (Nullable a)` and `Wire (Nullable (Nullable a))`.

Thus your first instance contains redundancy, since you can never receive `Just Nothing`.  Your second one looks like a generalisation of that.

My suggested fix (at least for the short term) is to avoid typeclasses, since they are very tricky, and just provide

```
joinNullable :: ExprArr (Wire (Nullable (Nullable a))) (Wire (Nullable a))
joinNullable = unsafeCoerce
```

This belongs in `Nullable.hs`.  I actually thought I'd already added this, but it seems not.

**Comment by tomjaguarpaw** [comment]
I've also just renoticed my comment

```
-- NB: We probably also need an instance
--
--       Default NullMaker (Wire (Maybe a)) (Wire (Maybe a))
```

I'm not sure this is the best way to solve the problem but it would help.

**Comment by tomjaguarpaw** [comment]
Did you come up with a satisfactory workaround in the meantime?

**Comment by bergmark** [comment]
https://github.com/silkapp/karamaan-opaleye/commit/65fc5e689a6185b20440bf46691dd3c6d3deacc1

**Comment by tomjaguarpaw** [comment]
Glad you've got something working.  A couple of comments:
1. I think it would be better to use `joinNullable` which I just added here: https://github.com/karamaan/karamaan-opaleye/commit/0b875092fa7f406074caab8068a87c04f414e568 (Admittedly this is not "automatic" like the `Default` mechanism).
2. If you want to keep going with the `Default QueryRunner` method, then your instance should be `... => Default QueryRunner (Wire (Maybe (Maybe a))) (Maybe a)` because you can never get `Just Nothing`.  (You may have to `fmap fromMaybe` to get this).

In the long run I should work out how to improve the `NullMaker` typeclass so that it doesn't turn `Nullable a` to `Nullable (Nullable a)` but keeps it at `Nullable a`.

**Comment by tomjaguarpaw** [comment]
It's easier than I thought.  Please try this in `LeftJoin.hs` and let me know how it works out

```
instance Default NullMaker (Wire (Maybe a)) (Wire (Maybe a)) where
  def = NullMaker id (Op2.constantLit PQ.NullLit)
```

**Comment by bergmark** [comment]
So should the NullMaker instance be the only change that I need? I seem to still need the `(FromField a, Default QueryRunner (Wire (Maybe a)) (Maybe a)) => Default QueryRunner (Wire (Maybe (Maybe a))) (Maybe (Maybe a))` instance.

**Comment by tomjaguarpaw** [comment]
The instance should be the only change.  You'll have to give a type signature to your left join so that it knows you want a `Nullable a` to come out, rather than a `Nullable (Nullable a)`.
